src/demoAskScriptServer.ts

```diff
@@ -0,0 +1,152 @@
+import bodyParser from 'body-parser';
+import cors from 'cors';
+import express from 'express';
+
+import { askCodeToSource } from './askcode';
+import {
+  resources as builtInResources,
+  runUntyped,
+  resource,
+  Resources,
+} from './askvm';
+import { parse as parseAskScript, AskScriptCode } from './askscript';
+
+import chalk = require('chalk');
+import { customAlphabet } from 'nanoid';
+
+const nanoid = customAlphabet('1234567890abcdef', 8);
+
+const values = {
+  clientNames: ['a', 'b', 'c', 'd', 'e', 'f', 'g'],
+  hello: 'Hi! This is a AskVM server running on localhost',
+  revPerClient: {
+    a: 426,
+    b: 35,
+    c: 446,
+    d: 246,
+    e: 133,
+    f: 136,
+    g: 53,
+  },
+
+  firstName: 'Luke',
+  middleName: 'LukeLuke',
+  lastName: 'Skywalker',
+  fullName: 'Luke Skywalker',
+  parents: [
+    {
+      firstName: 'Padmé',
+      lastName: 'Amidala',
+    },
+    {
+      firstName: 'Anakin',
+      lastName: 'Skywalker',
+    },
+  ],
+  friends: [
+    {
+      id: 0,
+      firstName: 'Padmé',
+      lastName: 'Amidala',
+    },
+    {
+      id: 1,
+      firstName: 'Anakin',
+      lastName: 'Skywalker',
+    },
+  ],
+};
+
+export const customResources: Resources = {
+  withSmile: resource<string, [string]>({
+    resolver: async (s: string): Promise<string> => {
+      return ':) ' + s + ' :)';
+    },
+  }),
+
+  hi: resource<string, []>({
+    resolver: async (): Promise<string> => {
+      return "Hi, this is AskQL server! It's " + new Date().toString();
+    },
+  }),
+};
+
+const baseEnvironment = {
+  resources: { ...builtInResources, ...customResources },
+  values,
+};
+
+const app = express();
+app.use(cors());
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: true }));
+
+app.get('/', async (req, res) => {
+  res.type('html');
+  res.send(
+    '<html><head><title>AskQL</title></head><body><p><b>Ask</b> me anything!</p><p><a href="http://askql.org/">AskQL website</a></p></body></html>'
+  );
+});
```

Comment:
if this is to serve as some human-readable documentation for the server it'd be good to extend it a little bit and maybe list those custom resources very simply?

Reply:
The resources are provided in Playground frontend as a link to the source code.